README.md
# NAME

Win32::Daemon - Extension enabling Win32 Perl scripts to run as a true Win32 service.

# SYNOPSIS

```perl
use Win32::Daemon;
Win32::Daemon::StartService();
# ...process Perl code...
Win32::Daemon::StopService();
```

# DESCRIPTION

This extension enables a Win32 Perl script to act as a true Win32 service.

# FUNCTIONS

## Function List

> `AcceptedControls()`
>
> `CallbackTimer()`
>
> `CreateService()`
>
> `ConfigureService()`
>
> `QueryServiceConfig()`
>
> `DeleteService()`
>
> `GetLastError()`
>
> `GetSecurity()`
>
> `GetServiceHandle()`
>
> `HideService()`
>
> `QueryLastMessage()`
>
> `RegisterCallbacks()`
>
> `RestoreService()`
>
> `SetSecurity()`
>
> `SetServiceBits()`
>
> `ShowService()`
>
> `StartService()`
>
> `State()`
>
> `StopService()`
>
> `Timeout()`

## Function Descriptions

- AcceptedControls( \[$NewControls\] )

    This function queries (and optionally sets) the current list of controls that the service registers for.
    By registering for a control the script is notifying the SCM that it is accepting the specified
    control messages. For example, if you specify the `SERVICE_ACCEPT_PAUSE_CONTINUE` control then
    the SCM knows that the script will accept and process any attempt to pause and continue (resume
    from paused state) the service.

    Recognized accepted controls:

    ```perl
    SERVICE_ACCEPT_STOP............The service accepts messages to stop.
    SERVICE_ACCEPT_PAUSE_CONTINUE..The service accepts messages to pause
                                   and continue.
    SERVICE_ACCEPT_SHUTDOWN........The service accepts messages to
                                   shutdown the system: when the OS is
                                   shutting down the service will be
                                   notified when it has accepted this
                                   control.
    ```

    Following controls are only recognized on Windows 2000 and higher:

    ```
    SERVICE_ACCEPT_PARAMCHANGE.....The service accepts messages
                                   notifying it of any parameter change
                                   made to the service.
    SERVICE_ACCEPT_NETBINDCHANGE...The service accepts messages
                                   notifying it of any network binding
                                   changes.
    ```

    By default all of these controls are accepted. To change this pass in a value consisting of
    any of these values OR'ed together.

    **NOTE:** You can query and set these controls at any time. However it is only supported to
    set them before you start the service (calling the `StartService()` function).

- CallbackTimer( \[ $NewTimerValue \] )

    This function returns the value of the callback timer. The value is in milliseconds.
    This value indicates how often the "Running" callback subroutine will be called. Note
    that the calling of this routine will be blocked by any other callback.

    If you pass in a value it will reset the timer to the specified frequency. Passing in
    a 0 will disable all "Running" callbacks. Passing in -1 will toggle the state between
    calling the "Running" callback subroutine and not calling it.

- CreateService ( \\%ServiceInfo )

    This function creates a new service in the system configuration. The
    return is TRUE if the service was created, and FALSE otherwise. If an error
    occurred, call GetLastError to retrieve the actual error code.

    **NOTE:** This function will fail if the script is not running with administrator
    privileges.

    The hash describes the service to be created. The keys are:

    - `name`

        The 'internal' service name; that is, the name of the
        registry key used to store the information on this service.

    - `display`

        The 'display' service name; that is, the name displayed
        by the services control panel or MMC plugin.

    - `path`

        The full path name to the executable. This should be the path to your Perl
        executable, which will normally be the contents of `$^X`.

        **NOTE:** If you are using a compiled perl script (such as one
        generated with PerlApp) as opposed to a text based perl script file then this
        value must point to the actual compiled script's executable (eg. `MyCompiledPerlService.exe`)
        instead of (`$^X` which usually points to `perl.exe`). You can specify
        any parameters to pass into the service using the `parameters` key.

    - `user`

        The username the service is to run under; this is optional.

    - `password`

        The password to be used to log in the service; this is
        technically optional, but needs to be specified if `user` is.

    - `parameters`

        The parameters to be passed to Perl; in other words, the command line you
        would execute interactively, but without the leading \`\`perl ''. The `parameters` key
        value is appended to the `path` key when starting the service.
        Typically this will be something like:

        ```
        MyPerlScript.pl /a /b /c
        ```

    - `machine`

        The name of the machine to create the service on. Omission
        or an empty string specify the machine executing the call.

    - `service_type`

        An integer representing the type of the service;
        defaults to `SERVICE_WIN32_OWN_PROCESS`.

    - `start_type`

        An integer specifying how (or whether) the service is
        to be started. The default is `SERVICE_AUTO_START`.

    - `error_control`

        An integer specifying how the Service Control Manager
        is to react if the service fails to start. The default is
        `SERVICE_ERROR_IGNORE`, which in fact gets you an error log entry.

    - `load_order`

        The name of the load order group of which this service
        is a member. The default is membership in no group. See value
        `ServiceGroupOrder` in registry key
        `HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control`
        for the names.

    - `tag_id`

        An integer representing the startup order of the service
        within its load ordering group.

    - `dependencies`

        A reference to the _internal_ names of services and/or
        load ordering groups upon which this service depends. The default is
        no dependencies. Load order group names are prefixed with a '+' to
        distinguish them from service names.

    - `description`

        A short text description of the service, displayed
        (at least) as flyover help by the MMC "services" plugin.

- ConfigureService( \\%ServiceInfo )

    Modify a service created with `CreateService`. Same arguments as
    `CreateService`.

    If you specify a `parameters` key you MUST specify a `path` key.

- DeleteService ($Machine, $ServiceName )

    This function deletes an existing service. The return is TRUE if the
    service was deleted, and FALSE otherwise. If an error occurred, call
    GetLastError to retrieve the actual error code.

    The arguments are the name of the machine (an empty string specifies
    the machine executing the call), and the 'internal' service name (i.e.
    the string passed in the `name` element when the service was created).

    A running service may not be deleted.

- GetSecurity( $Machine, $ServiceName )

    This will return a binary Security Descriptor (SD) that is associated with the
    specified service on the specified machine.

    The SD is in self-relative format. It can be imported into a `[Win32::Perms](https://metacpan.org/pod/Win32%3A%3APerms)` object using
    the `Win32::Perms` object's `Import()` method.

- RegisterCallbacks( $CodeRef | \\%Hash )

    This will register specified code subroutines that will be called when specified
    events take place. For example if you register a subroutine with the
    `pause` event then this routine will be called when there is an attempt to pause the
    service. Not all events must have callbacks registered.

    If only a reference to a subroutine is passed in then it will be called for each and every
    event. You can pass in a hash containing particular key names (listed below) with
    code references.

    Possible hash key names:

    ```perl
    Key Name                 Event
    -------------            --------------------------------------
    start....................The service is starting.
    pause....................The service is entering a paused state.
    continue.................The service is resuming from a paused
                             state.
    stop.....................The service is stopping (see note below).
    running..................The service is running (see note below).
    interrogate..............The service is being queried for
                             information.
    shutdown.................The system is being shut down.
    preshutdown..............The system is about to begin shutting
                             down (Vista+ only).
    param_change.............There has been a parameter change to
                             the system.
    net_bind_add.............A new network binding has been made.
    net_bind_remove..........A network binding has been removed.
    net_bind_enable..........A network binding has been enabled.
    net_bind_disable.........A network binding has been disabled.
    device_event.............A device has generated some event.
    hardware_profile_change..A change has been made to the system's
                             hardware profile.
    power_event..............A power event has occured (eg change to
                             battery power).
    session_change...........There has been a change in session.
    user_defined.............A user defined event has been sent to
                             the service.
    ```

    **NOTES:**

    - The `Stop` state

        When a service calls into the registered "stop" callback routine
        the script should call the `StopService()` function. This tells the service to terminate
        and return back to the Perl script. This is the only way for the service to know that it
        must stop.

    - The `Running` state

        Periodically the extension will call into a registered
        "Running" subroutine. This allows the script to process data. This routine should be fast
        and return quickly otherwise it will block other callback events from being run. The
        frequency of calling the "Running" subroutine is dictated by the callback timer value
        passed into `StartService()` and any changes made to this value by calling into
        `CallbackTimer()`.

- SetSecurity( $Machine, $ServiceName, $BinarySD | $Win32PermsObject )

    This applies the specified Security Descriptor (SD) to the specified service on the
    specified machine. You must have appropriate permissions to call this function.

    The specified SD can be either a binary SD (in self-relative or absolute format) or
    it can be a Win32::Perms object.

    This only sets the DACL and SACL. The owner and group are not set even if they are
    specified in the SD.

- StartService( \[ \\%Context, $CallbackTimer \] )

    This starts a new service thread. The script should call this as soon as possible.  When
    the service manager starts the service Perl is started and the script is loaded.

    This function returns the thread handle of the service thread. If you call into this more
    than once it will only return the thread handle (it won't create another new service thread).

    - Callback Mode

        If the script has already registered callback routines (using `RegisterCallbacks()`) then
        the call into `StartService()` will not return until the service has stopped. However
        callbacks will be made for each state change and callback timer timeout (refer to `RegisterCallbacks()`).

- StopService()

    This will instruct the service to terminate.

- Timeout( \[$TimeoutValue\] )

    This function sets the new timeout value indicating how long a command will wait before
    Win32::Daemon tells the Service Control Manager that the command failed.

- QueryLastMessage( \[$fResetMessage\] )

    This function returns the last message that the service manager has sent to the service.

    Pass in a non zero value to reset the pending message to `SERVICE_CONTROL_NONE`. This way
    your script can tell when two of the same messages come in.

    Occasionally the service manager will send messages to the service. These messages
    typically request the service to change from one state to another.  It is important that
    the Perl script responds to each message otherwise the service manager becomes confused
    about the current state of the service. For example, if the service manager is submits
    a `SERVICE_PAUSE_PENDING` then it expects the Perl script to recognize the change to a paused
    state and submit the new state by calling `State( SERVICE_PAUSED )`.

    You can update the service manager with the current status using the `State()` function.

    Possible values returned are:

    ```perl
    Valid Service Control Messages
    ------------------------------
    SERVICE_CONTROL_NONE............No message is pending.
    SERVICE_CONTROL_STOP............The SCM is requesting the service to
                                    stop. This results in State()
                                    reporting SERVICE_STOP_PENDING.
    SERVICE_CONTROL_PAUSE...........The SCM is requesting the service to
                                    pause. This results in State()
                                    reporting SERVICE_PAUSE_PENDING.
    SERVICE_CONTROL_CONTINUE........The SCM is requesting the service to
                                    continue from a paused state. This
                                    results in State() reporting
                                    SERVICE_CONTINUE_PENDING.
    SERVICE_CONTROL_INTERROGATE.....The service manager is querying the
                                    service's state

    SERVICE_CONTROL_USER_DEFINED....This is a user defined control.
                                    There are 127 of these beginning
                                    with SERVICE_CONTROL_USER_DEFINED
                                    as the base.
    ```

    Windows 2000 specific messages:

    ```perl
    SERVICE_CONTROL_SHUTDOWN........The machine is shutting down. This
                                    indicates that the service has
                                    roughly 20 seconds to clean up and
                                    terminate. This time can be extended
                                    by submitting SERVICE_STOP_PENDING
                                    via the State() function.
    SERVICE_CONTROL_PARAMCHANGE.....Service parameters have been
                                    modified.
    SERVICE_CONTROL_NETBINDADD......A network binding as been added.
    SERVICE_CONTROL_NETBINDREMOVE...A network binding has been removed.
    SERVICE_CONTROL_NETBINDENABLE...A network binding has been enabled.
    SERVICE_CONTROL_NETBINDDISABLE..A network binding has been disabled.
    SERVICE_CONTROL_DEVICEEVENT.....A device has generated some event.
    SERVICE_CONTROL_HARDWAREPROFILECHANGE
                                    A change has been made to the
                                    system's hardware profile.
    SERVICE_CONTROL_POWEREVENT......A power event has occured (eg change
                                    to battery power).
    SERVICE_CONTROL_SESSIONCHANGE...There has been a change in session.
    ```

    Windows Vista+ specific messages:

    ```
    SERVICE_CONTROL_PRESHUTDOWN ....The machine is about to shut down.
                                    This provides the service much more
                                    time to shutdown than
                                    SERVICE_CONTROL_SHUTDOWN.
    ```

    **Note:** When the system shuts down it will send a `SERVICE_CONTROL_SHUTDOWN` message. The
    Perl script has approximately 20 seconds to perform any shutdown activities before the
    Control Manger stops the service. If more time is needed call the `State()` function
    passing in the `SERVICE_STOP_PENDING` control message along with how many seconds it will
    take to shutdown the service. This time value is only an estimate. When the service is
    finally ready to stop it must submit the `SERVICE_STOPPED` message as in:

    ```
    if( SERVICE_CONTROL_SHUTDOWN == State() )
    {
        Win32::Daemon::State( SERVICE_STOP_PENDING, 30_000 );
        #...process code...
        Win32::Daemon::State( SERVICE_STOPPED );
    }
    ```

- State(\[$NewState \[, $Hint \] || \\%Hash \] )

    This function returns the current state of the service.  It can optionally update the
    status of the service as well.  This is the last status reported to the service manager.

    Optionally you can pass in a value that will be sent to the service manager.
    Optionally you can pass in a numeric value indicating the "hint". This is the number of
    milliseconds the SCM can expect to wait before the service responds to the request. For example,
    if your service script reports a hint of 30,000 milliseconds means that the SCM will have to wait
    for 30 seconds for the script to change the service's state before deciding that the
    script is non responsive.

    If you are setting/updating the state instead of passing in the state and wait hint you could
    pass in a hash reference. This allows you to specify the state, wait hint and error state. You
    can use the following keys:

    ```
    Hash Key
    --------
    state..........Valid service state (see table below).
    waithint.......A wait hint explained above. This is in milliseconds.
    error..........Any 32 bit error code. This is what will be reported
                   if an application queries the error state of the
                   service. It is also what is reported if a call to
                   start the services fails.
                   To reset an error state pass in NO_ERROR.
                   The only invalid error value is 0xFFFFFFFF.
    ```

    Example of passing in an error:

    ```perl
    Win32::Daemon::State( { error => 0x12345678 } );
    # Later to reset the error:
    Win32::Daemon::State( { error => NO_ERROR } );
    ```

    Possible values returned (or submitted):

    ```perl
    Valid Service States
    --------------------
    SERVICE_NOT_READY..........The SCM has not yet been initialized. If
                               the SCM is slow or busy then this value
                               will result from a call to State().
                               If you get this value, just keep calling
                               State() until you get
                               SERVICE_START_PENDING.
    SERVICE_STOPPED............The service is stopped.
    SERVICE_RUNNING............The service is running.
    SERVICE_PAUSED.............The service is paused.
    SERVICE_START_PENDING......The service manager is attempting to
                               start the service.
    SERVICE_STOP_PENDING.......The service manager is attempting to
                               stop the service.
    SERVICE_CONTINUE_PENDING...The service manager is attempting to
                               resume the service.
    SERVICE_PAUSE_PENDING......The service manager is attempting to
                               pause the service.
    ```

# Callbacks

Callbacks were introduced in version v20030617.

The Win32::Daemon supports the concept of event callbacks. This allows a script to
register a particular subroutine with a particular event. When the event occurs it
will call the Perl subroutine registered with that event. This can make it very simple
to write scripts.

You register a callback subroutine by calling into the `RegisterCallbacks()` function.
You can pass in a code reference or a hash. A code reference will register the specified
subroutine with all events. A hash allows you to pick which events you want to
register for which subroutines. You do not have to register all events. If an event is
not registered for a subroutine then the script will not be notified when the event
occurs.

At a minimum a script should register for the 'Start' and 'Running' states. This enables
the script to actually start and to periodically process data.

When an event callback occurs the subroutine should change the state accordingly by
passing in the new state into `State()`. For example the 'Start' callback would call
`State( SERVICE_RUNNING )` to inform the service that it is officially running. Another
example is the 'Pause' state should call `State( SERVICE_PAUSED )` to inform the service
that it is officially paused.

Once callback subroutines are registered the script enters the service mode by calling
`StartService()`. This will being the process of calling the event callback routines.
Note that when callback routines are registered the `StartService()` function will not
return until a callback routine calls `StopService()` (typically the 'Stop' event callback
would call `StopService()`.

When calling into `StartService()` you can pass in a hash reference. This reference is known as
a "context" hash. For every callback the hash will be passed into the callback routine. This enables
a script to query and set data in the hash--essentially letting you pass information across to
different callback events. This context hash is not required.

When a callback is made it always passes two parameters in: $State and $Context. $State is simply
the state change that caused the callback. This represents the event that took place (e.g. `SERVICE_PAUSE_PENDING`,
`SERVICE_START_PENDING`, etc). The $Context is a reference to the context hash that was passed into
the `StartService()` function.

A typical callback routine should look similar to:

```perl
sub Callback_Start
{
    my( $Event, $Context ) = @_;
    $Context->{last_event} = $Event;

    # ...do some work here...

    # Tell the service manager that we have now
    # entered the running state.
    Win32::Daemon::State( SERVICE_RUNNING );
    return();
}
```

Refer to `Example 4: Using a single callback` and `Example 5: Using different callback routines` for an example of using callbacks.

# COMPILED PERL APPLICATIONS

Many users like to compile their perl scripts into executable programs. This way it is much easier to copy them around
from machine to machine since all necessary files, packages and binaries are compiled into one .exe file. These compiled
perl scripts are compatible with Win32::Deamon as long as you install it correctly.

If you are going to compile your Win32::Daemon based perl script into an .exe there is nothing unique you need to do
to your Win32::Daemon code with one single exception of the call into Win32::Daemon::`CreateService()`. When passing in
the 'path' and 'parameters' values into `CreateService()` observe the following simple rules:

- If using a Perl script

    ```perl
    path........The full path to the Perl interpeter ($^X).
                This is typically:
                  c:\perl\bin\perl.exe

    parameters..This value MUST start with the full path to the perl
                script file and append any parameters
                that you want passed into the service. For
                Example:
                  c:\scripts\myPerlService.pl -param1 -param2 "c:\\Param2Path"
    ```

- If using a compiled Perl application

    ```perl
    path........The full path to the compiled Perl application.
                For example:
                  c:\compiledscripts\myPerlService.exe

    parameters..This value is just the list of  parameters
                that you want passed into the service. For
                Example:
                  -param1 -param2 "c:\\Param2Path"
    ```

Refer to ["Example 3: Install the service"](#example-3-install-the-service) for an example.

# EXAMPLES

## Example 1: Simple Service

This example service will delete all .tmp files from the c:\\temp directory every
time it starts.  It will immediately terminate.

```perl
use Win32::Daemon;

# Tell the OS to start processing the service...
Win32::Daemon::StartService();

# Wait until the service manager is ready for us to continue...
while( SERVICE_START_PENDING != Win32::Daemon::State() )
{
    sleep( 1 );
}

# Now let the service manager know that we are running...
Win32::Daemon::State( SERVICE_RUNNING );

# Okay, go ahead and process stuff...
unlink( glob( "c:\\temp\\*.tmp" ) );

# Tell the OS that the service is terminating...
Win32::Daemon::StopService();
```


Changes
Revision history for Perl extension {{$dist->name}}

{{$NEXT}}
    - Fix a typo in the docs. Thanks, teridon. (GH PR#18)

20200728  2020-07-28
    - dwControlsAccepted wasn't being initialized properly on Windows 10
      (#14, #16)

20190315  2019-03-16
    - fix and simplify constant exporting (#13)

20181025  2018-10-25
    - Fix the build by fixing the case sensitive Constant to constant
      (RT #126272 and RT #75506)
    - Cleaned up and organized this change log
    - Use Dist::Zilla to prepare the releases
    - Maintain dependencies in cpanfile
    - Test with appveyor
    - Move Daemon.pm to its proper spot in the lib/Win32 directory
    - Take Cygwin out of the supported OSes as there are many compilation errors
      at the moment. Please submit a PR if you are capable of fixing this!
    - Document that CreateService() requires admin rights and don't hide
      "permission denied" error (RT #73715, RT #86501 and RT #86882)

20131206  2013-12-06
    - Fix required perl version 5.6 -> 5.006
    - Add Github repo link to META.yml
    - Typo fixes by David Steinbrunner

20120725  2012-07-25
    - POD cleanup
    - Doc fix from Brian Wightman (github #1, #2): wait hint in service
      service status is in ms

20110117  2011-10-17
    - make sure DllMain() has "C" linkage when compiled with MinGW
      (Haiko Strotbek). [rt#64575]
    - change StartService() and GetServiceHandle() to return proper
      64-bit value on Win64 (Haiko Strotbek). [rt#64577]

20101014  2010-10-14
    - minimal POD cleanup
    - avoid CPAN indexing of the packages under test/
    - QueryLastMessage() had been renamed to LastControlMessage().
      Rename it back to preserve backwards compatibility.

20091028  2009-10-28
    - Disabled END {} in non-main threads to fix bug RT#50020 (Olivier Mengué)

20080321  2008-03-21
    - Added support for SERVICE_CONTROL_PRESHUTDOWN.
    - Callbacks can now return a state value. This is the functional equivalent
      of calling Win32::Daemon::State( $NewState ) before returning:
        //	sub EventCallback
        //	{
        //	  my( $Event, $Context ) = @_;
        //		# Process the event
        //		Win32::Daemon::State( $NewState );
        //	  return;
        //	}
        //	=================== SAME AS ====================
        //	sub EventCallback
        //	{
        //	  my( $Event, $Context ) = @_;
        //		# Process the event
        //	  return( $NewState );
        //	}
    - Callback timer works correctly!
    - CALLBACK_RUNNING has been depreciated and replaced with CALLBACK_TIMER.
      It is only called when the timer (set by CallbackTimer()) has expired.
      Use this as a heartbeat. You only need to set it once either with
      CallbackTimer() or passing in the timer value when calling into
      StartService(). Set the callback using "timer" as the callback name.
      Using "running" will also work but it is mapped to "timer". If you
      specify both, only "timer" will be registered.
    - Created TestFramework.pm to help with testing services.
    - Pruned out unnecessary stuff from Daemon.pm file.

20070102  2007-01-02
    - Cleaned up a bit.
    - Added WM_QUIT message to the ServiceMain function. Now the Perl StopService() will
      post this message to shut down the service thread.
    - Fixed bug where messages were posted to wrong thread.

20061222  2006-12-22
    - Converted to XS file.
    - Fixed callback heartbeat: now properly calls back with SERVICE_RUNNING
      (not SERVICE_CONTROL_RUNNING)
    - StopService() will post WM_QUIT message to the ServiceMain() thread to
      shut down the service thread.
    - Calling into StopService() will auto change the state to STOPPING/STOPPED
      so you do not need to explicitly do so (calling State() or a callback
      returning STOPPING/STOPPED).

20030617  2003-06-17
    - Added support for querying and setting service security. This includes
      DACL and SACL support. Due to complexity and failures you can not set
      owner or group. But in most cases (99%) this is not needed.
    - GetSecurity()
    - SetSecurity()
    - Finished incorporating callback routines.
    - Added a few more events.

20020605  2002-06-05
    - Added support for reporting service errors. You can now pass in a
      hash reference into State(). More details in the POD docs.
    - Modified: UpdateServiceStatus() function to accept a 3rd parameter
      (dwError). This allows the calling code to report a service error.

20020114  2002-01-14
    - Fixed another bug where service wouldn't work on WinXP machines. Variable
      recognized_controls was wrongly initialized for WinXP systems. This
      rendered the service unable to start, pause or stop. Fix was only in
      the Daemon.pm file. (marc.pijnappels@nec-computers.com)

20020108  2002-01-08
    - Fixed another bug wwhere servie wouldn't work on Win2k machines. We
      were treating Win2k (version 5.0) the same as WinXP (version 5.1) and
      adding WinXP specific controls. This rendered the service unable to
      start, pause or stop. Fix was only in the Daemon.pm file.

20011230  2001-12-30
    - Fixed bug where service doesn't work properly with Windows NT 4. We were
      defaulting by acccepting the SERVICE_ACCEPT_PARAMCHANGE and
      SERVICE_ACCEPT_NETBINDCHANGE controls. However, they were introduced in
      Win2k so NT 4 coughed up blood with them. (rothd@roth.net)

20011221  2001-12-21
    - Fixed bug where service doesn't work properly with Windows NT 4. We were
      defaulting by acccepting the SERVICE_ACCEPT_PARAMCHANGE and
      SERVICE_ACCEPT_NETBINDCHANGE controls. However, they were introduced in
      Win2k so NT 4 coughed up blood with them.

20011217  2001-12-17
    - Added support for .Net Server accept control contants:
      SERVICE_ACCEPT_HARDWAREPROFILECHANGE
      SERVICE_ACCEPT_POWEREVENT
      SERVICE_ACCEPT_SESSIONCHANGE

20011205  2001-12-05
    - AcceptedControls()

constant.cpp
//////////////////////////////////////////////////////////////////////////////
//
//  Constant.cpp
//  Win32::Daemon Perl extension constants source file
//
//  Copyright (c) 1998-2008 Dave Roth
//  Courtesy of Roth Consulting
//  http://www.roth.net/
//
//  This file may be copied or modified only under the terms of either
//  the Artistic License or the GNU General Public License, which may
//  be found in the Perl 5.0 source kit.
//
//  2008.03.24  :Date
//  20080324    :Version
//////////////////////////////////////////////////////////////////////////////

#define WIN32_LEAN_AND_MEAN

#ifdef __BORLANDC__
typedef wchar_t wctype_t; /* in tchar.h, but unavailable unless _UNICODE */
#endif

#include <windows.h>
#include <tchar.h>
#include <wtypes.h>
#include <stdio.h>      //  Gurusamy's right, Borland is brain damaged!
#include <math.h>       //  Gurusamy's right, MS is brain damaged!

#include <winspool.h>
#include <LMACCESS.H>
#include <LM.H>
#include <LMAUDIT.H>
#include <LMERR.H>
#include <LMERRLOG.H>

#include "constant.h"

/* needed for compatibility with perls 5.14 and older */
#ifndef newCONSTSUB_flags
#define newCONSTSUB_flags(stash, name, len, flags, sv) newCONSTSUB((stash), (name), (sv))
#endif

#define DO_CONST_IV(c) \
    newCONSTSUB_flags(stash, #c, strlen(#c), 0, newSViv(c)); \
    av_push(export_av, newSVpvs(#c));
#define DO_CONST_PV(c) \
    newCONSTSUB_flags(stash, #c, strlen(#c), 0, newSVpvn(c, strlen(c))); \
    av_push(export_av, newSVpvs(#c));

// SC_GROUP_IDENTIFIER is a char constant, we need to make it a string
// before we pass it to DO_CONST_PV()
static const char sc_group_identifier_str[] = {SC_GROUP_IDENTIFIER, 0};

void ExportConstants(pTHX)
{
    HV *stash = gv_stashpv("Win32::Daemon", GV_ADD);
    AV *export_av = get_av("Win32::Daemon::EXPORT_XS", TRUE);

    DO_CONST_IV(SERVICE_NOT_READY);
    DO_CONST_IV(SERVICE_STOPPED);
    DO_CONST_IV(SERVICE_RUNNING);
    DO_CONST_IV(SERVICE_PAUSED);
    DO_CONST_IV(SERVICE_START_PENDING);
    DO_CONST_IV(SERVICE_STOP_PENDING);
    DO_CONST_IV(SERVICE_CONTINUE_PENDING);
    DO_CONST_IV(SERVICE_PAUSE_PENDING);

    DO_CONST_IV(SERVICE_CONTROL_NONE);
    DO_CONST_IV(SERVICE_CONTROL_STOP);
    DO_CONST_IV(SERVICE_CONTROL_PAUSE);
    DO_CONST_IV(SERVICE_CONTROL_CONTINUE);
    DO_CONST_IV(SERVICE_CONTROL_INTERROGATE);
    DO_CONST_IV(SERVICE_CONTROL_SHUTDOWN);
    DO_CONST_IV(SERVICE_CONTROL_PARAMCHANGE);
    DO_CONST_IV(SERVICE_CONTROL_NETBINDADD);
    DO_CONST_IV(SERVICE_CONTROL_NETBINDREMOVE);
    DO_CONST_IV(SERVICE_CONTROL_NETBINDENABLE);
    DO_CONST_IV(SERVICE_CONTROL_NETBINDDISABLE);
    DO_CONST_IV(SERVICE_CONTROL_DEVICEEVENT);
    DO_CONST_IV(SERVICE_CONTROL_HARDWAREPROFILECHANGE);
    DO_CONST_IV(SERVICE_CONTROL_POWEREVENT);
    DO_CONST_IV(SERVICE_CONTROL_SESSIONCHANGE);
    DO_CONST_IV(SERVICE_CONTROL_USER_DEFINED);
    DO_CONST_IV(SERVICE_CONTROL_RUNNING);
    DO_CONST_IV(SERVICE_CONTROL_TIMER);
    DO_CONST_IV(SERVICE_CONTROL_START);

#ifdef SERVICE_CONTROL_PRESHUTDOWN
    DO_CONST_IV(SERVICE_CONTROL_PRESHUTDOWN);
#endif

    //  Service bits available to a script
    DO_CONST_IV(USER_SERVICE_BITS_1);
    DO_CONST_IV(USER_SERVICE_BITS_2);
    DO_CONST_IV(USER_SERVICE_BITS_3);
    DO_CONST_IV(USER_SERVICE_BITS_4);
    DO_CONST_IV(USER_SERVICE_BITS_5);
    DO_CONST_IV(USER_SERVICE_BITS_6);
    DO_CONST_IV(USER_SERVICE_BITS_7);
    DO_CONST_IV(USER_SERVICE_BITS_8);
    DO_CONST_IV(USER_SERVICE_BITS_9);
    DO_CONST_IV(USER_SERVICE_BITS_10);

    //  Define Service Types
    DO_CONST_IV(SERVICE_WIN32_OWN_PROCESS);
    DO_CONST_IV(SERVICE_WIN32_SHARE_PROCESS);
    DO_CONST_IV(SERVICE_KERNEL_DRIVER);
    DO_CONST_IV(SERVICE_FILE_SYSTEM_DRIVER);
    DO_CONST_IV(SERVICE_INTERACTIVE_PROCESS);

    //    Define control acceptance constants
    DO_CONST_IV(SERVICE_ACCEPT_STOP);
    DO_CONST_IV(SERVICE_ACCEPT_PAUSE_CONTINUE);
    DO_CONST_IV(SERVICE_ACCEPT_SHUTDOWN);
    DO_CONST_IV(SERVICE_ACCEPT_PARAMCHANGE);
    DO_CONST_IV(SERVICE_ACCEPT_NETBINDCHANGE);

#ifdef SERVICE_ACCEPT_HARDWAREPROFILECHANGE
    DO_CONST_IV(SERVICE_ACCEPT_HARDWAREPROFILECHANGE);
#endif // SERVICE_ACCEPT_HARDWAREPROFILECHANGE

#ifdef SERVICE_ACCEPT_POWEREVENT
    DO_CONST_IV(SERVICE_ACCEPT_POWEREVENT);
#endif // SERVICE_ACCEPT_POWEREVENT

#ifdef SERVICE_ACCEPT_SESSIONCHANGE
    DO_CONST_IV(SERVICE_ACCEPT_SESSIONCHANGE);
#endif // SERVICE_ACCEPT_SESSIONCHANGE

    //  Define Start Types
    DO_CONST_IV(SERVICE_BOOT_START);
    DO_CONST_IV(SERVICE_SYSTEM_START);
    DO_CONST_IV(SERVICE_AUTO_START);
    DO_CONST_IV(SERVICE_DEMAND_START);
    DO_CONST_IV(SERVICE_DISABLED);

    //  Define Error Controls
    DO_CONST_IV(SERVICE_ERROR_NORMAL);
    DO_CONST_IV(SERVICE_ERROR_SEVERE);
    DO_CONST_IV(SERVICE_ERROR_CRITICAL);

    // Define the Group Identifier (prepend this value to the name of a dependent group)
    DO_CONST_PV(sc_group_identifier_str);

    // Define the state's default error
    DO_CONST_IV(NO_ERROR);
};

/*
HISTORY

   20020605 rothd
    - Added the NO_ERROR constant.


    20080321 rothd
        -Added SERVICE_CONTROL_PRESHUTDOWN.
        -Added SERVICE_CONTROL_TIMER
        -Added SERVICE_CONTROL_START
        -Fixed constant look up to properly manage strings
*/
